### PR TITLE
Family hierarchy CPEX0043

### PR DESCRIPTION
**Pull request by MicK7**

### CODE REVIEW

File: src/cgns_header.h [hunk 1]
```diff
@@ -31,6 +31,7 @@ typedef char char_66[66]; /* 32 + '/' + 32 + '\0' */
 #else
 typedef char char_66[33]; /* 32 + '\0' (caller's malloc compat issues) */
 #endif
+typedef char char_md[CG_MAX_GOTO_DEPTH*33]; /* ** FAMILY TREE ** */
```
Comment (poinot) [inline src/cgns_header.h]: Should be / + 32 * MAX GOTO + ending / + \0 which makes 33*MAX_GOTO + 2

File: src/cgnslib.c [hunk 2]
```diff
@@ -1325,63 +1324,147 @@ int cg_family_read(int file_number, int B, int F, char *family_name,
...
+
+        (*F) = index+1;
+
+        if( ! skip ) { /* If not an existing intermediate family node */
+            memset( family, 0, sizeof(cgns_family) );
+            strcpy( family->name, tok );
+
+            /* Save data in file */
+            if( cgi_new_node( parent_id, tok, "Family_t", &family->id, "MT", 0, 0, 0) ) {
+                return CG_ERROR;
```
Comment (poinot) [inline src/cgnslib.c]: Probably we should change all theses "*_t" magic values by pre-processing constants one of these days...

File: src/cgnslib.c [hunk 3]
```diff
@@ -1434,8 +1517,13 @@ int cg_family_name_write(int file_number, int B, int F,
     cgns_famname *famname = 0;
 
      /* verify input */
-    if (cgi_check_strlen(name) ||
-        cgi_check_strlen(family)) return CG_ERROR;
+    if (cgi_check_strlen(name) /*||
+        cgi_check_strlen(family)*/) return CG_ERROR;
+
+    if ( strlen(family) > 33*20 ) {
```
Comment (poinot) [inline src/cgnslib.c]: Constants required, whet is 33, what is 20 ?

File: src/tests/test_family_tree.c [hunk 4]
```diff
@@ -0,0 +1,446 @@
...
+
+int main ()
+{
+    int i, j, k, n, nfam, nb, ng, nnames, ier;
+
+    int cgfile, cgbase, cgtree, cgzone, cgfam, cgcoord, cgbc, cgsr;
+    float exp[5];
+    char outfile[33], name[33], tname[33];
+    char family_name[20*33];
+    char tfamily_name[20*33];
```
Comment (poinot) [inline src/tests/test_family_tree.c]: Constants again...

File: src/tests/test_family_tree.c [hunk 5]
```diff
@@ -0,0 +1,446 @@
...
+        cg_exponents_write(CGNS_ENUMV(RealSingle), exp) ||
+        cg_gopath(cgfile, "../CoordinateY") ||
+        cg_exponents_write(CGNS_ENUMV(RealSingle), exp) ||
+        cg_gopath(cgfile, "../CoordinateZ") ||
+        cg_exponents_write(CGNS_ENUMV(RealSingle), exp))
+        error_exit("coordinate write");
+
+    /* Write Family Tree */
+    /*
+     * FamilyTree
```
Comment (poinot) [inline src/tests/test_family_tree.c]: Probably we need a limits test with (1) empty strings  (2) only / chars (3) max number of chars

File: src/cgns_header.h (hunk 1 above)
Comment (MicK7) [reply src/cgns_header.h]: done

File: src/cgnslib.c (hunk 2 above)
Comment (MicK7) [reply src/cgnslib.c]: Let's do it in a separate branch to not infect this topic

File: src/cgnslib.c (hunk 3 above)
Comment (MicK7) [reply src/cgnslib.c]: should we have a cgi_check_tree_strlen ?

File: src/tests/test_family_tree.c (hunk 4 above)
Comment (MicK7) [reply src/tests/test_family_tree.c]: Of course

File: src/tests/test_family_tree.c (hunk 5 above)
Comment (MicK7) [reply src/tests/test_family_tree.c]: Should we add:
empty string --> check that it fails
only '/' --> check that it fails
max _number_of_char  --> pass
max _number_of_char +1 --> fail
??

Comment (MicK7) [reply src/tests/test_family_tree.c]: I added the tests

**Comment by MicK7** [comment]
@oma-numeca can you check that the final code agree with your intent ?

**Comment by MicK7** [comment]
@brtnfld can this topic be merged before you introduce the fix for "cgio_read_*_data_type"  ?

**Comment by brtnfld** [comment]
I don't see any overlapping changes, so the merges should be independent.

**Comment by MicK7** [comment]
@poinot I replaced 33 by CGIO_MAX_NAME_LENGTH+1 . It does the job properly.
@brtnfld Thanks for pointing out the missing fortran wrapping tests.

**Comment by MicK7** [comment]
I don't mind the squashing.

Concerning the testing, do you think it would be of interest to setup codacy on this github repository ( https://www.codacy.com or https://app.codacy.com/ ). It is free for public repository if only the maintainer account is linked (As far as I have seen other project use it on github). This would help in validating code quality before merging a branch and as well as detecting bad practice in the CGNS code base (duplication ...).

**Comment by oma-numeca** [comment]
> @oma-numeca can you check that the final code agree with your intent ?

Hi! No remarks on my side about the changes, all necessary features are there for our purposes.

**Comment by brtnfld** [comment]
I don't have any objections to using codacy.
